Creating new command for self-diagnosis

ni-reis opened the pull request:
PR for new ACR command for self-diagnosis `az acr check-health`.

More information can be found here:
https://gist.github.com/ni-reis/a9ef51a6f08a2694482be401b068e558

Examples of the command:

<img width="347" alt="img1" src="https://user-images.githubusercontent.com/44070983/58229609-8d93e100-7d08-11e9-8901-ac32b7690b93.PNG">
<img width="725" alt="img2" src="https://user-images.githubusercontent.com/44070983/58229611-8e2c7780-7d08-11e9-91b6-f4dc63775bfe.PNG">
<img width="1101" alt="img3" src="https://user-images.githubusercontent.com/44070983/58229613-8e2c7780-7d08-11e9-8e0d-cbda6c0ef726.PNG">
<img width="701" alt="img4" src="https://user-images.githubusercontent.com/44070983/58229615-8f5da480-7d08-11e9-9f53-1e705663c056.PNG">



---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

Commit:
Creating az acr check-health command

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,420 @@
+from subprocess import getoutput
+from knack.util import CLIError
+from .custom import get_docker_command
+from .helm import _get_helm_command
```
Could you also remove the underscore _ from `_get_helm_command`

ni-reis replied:
sure, will change it.

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,420 @@
+    else:
+        confirmation = ''
+
+    while confirmation not in ['Y', 'y', 'N', 'n']:
```
use `from knack.prompting import prompt_y_n` for confirmation

ni-reis replied:
oh, great that we have a method for this ;) will include in next commit

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py:
```diff
@@ -214,3 +214,12 @@ def load_arguments(self, _):  # pylint: disable=too-many-statements
+        c.argument('pull', options_list=['--pull'], help='Tries to pull the sample image mcr.microsoft.com/mcr/hello-world', action='store_true', required=False)
+        c.argument('all_checks', options_list=['-a', '--all-checks'], help='Gets all health reports supported.', action='store_true', required=False)
+        c.argument('skip_confirmation', options_list=['-y'], help='Skips confirmation, allowing to automatically pull the sample image', action='store_true', required=False)
+        c.argument('ignore_errors', options_list=['--ignore'], help='Ignore errors, displaying them only at the final', action='store_true', required=False)
```
IMO, too many flags would make it confusing to use. I initially thought just a command without any flags so that anyone can run it easily with a full report, not even the registry name since we can just pick one from the subscription (but error out if there is no registry).

Keep in mind that adding params is easy but removing them in the future is hard, so I would suggest adding the ones that are absolutely necessary and slowly adding new ones depending on feedback.

ni-reis replied:
Fair enough. Going by now with `'skip_confirmation'` and maybe `--ignore`

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,420 @@
+        "Recommendations : {}" \
+        .format(operation, error.name, recommendations_map[error.name])
+
+    if additional_message is not None:
```
if additional_message:

ni-reis replied:
acked, will change in next commit

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,420 @@
+def print_pass(message):
+    from colorama import Fore, Style, init
+    init()
+    print(str(message) + " : " + Fore.GREEN + "OK" + Style.RESET_ALL)
```
for the use of `print`, it's generally not allowed at each module level, unless there is a strong reason (e.g., build logs). Could you check if we can just use logger?
```
from knack.log import get_logger
logger = get_logger(__name__)
logger.warning
logger.error
...
```

ni-reis replied:
I tried to use the logger, but it didn't work as smoothly as the print (I didn't find a way to print it to stdout, which is the intention). Also I saw some other extensions using `print` in similar situations on other extensions

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,420 @@
+    # If there's no docker command there's no reason to proceed the operations
+    try:
+        docker_command = get_docker_command()
+    except CLIError:
```
There are multiple cases where CLIError may throw, docker not installed, docker daemon not running, etc. We might need some refactoring here with `get_docker_command` to capture the error more precisely.
It's okay for now to just rely on the existing logic.

ni-reis replied:
Agreed. We have to make `get_docker_command` errors more clear for the users.

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,420 @@
+    docker_version = ""
+
+    for line in lines:
+        if line.startswith("Server Version"):
```
can you use `docker -v` if you just need version? and then there is no need to parse the result since it's just one line, unless you plan to parse more info from `docker system info`.

ni-reis replied:
The reason for using `docker system info` is because `docker -v` doesn't catch errors with docker daemon and is always available, while `docker system info` catches an error with docker daemon.

djyou replied:
`_get_docker_command` already checked docker daemon status, that's why I was suggesting, either refactor the code, or  cleanup extra checks here.

ni-reis replied:
Oh, I see. I didn't know that getting docker command validated docker daemon already. Ok, will migrate to `docker -v` in the next commit

ni-reis replied:
Since `get_docker_command` already checks docker daemon status, I'll just print docker daemon as available if docker command is successfully retrieved. But I think it might be better to stick with docker system info by now in case we want to track other things in the future. This is just a first version of the command, and people will be free to add other things that they might consider necessary, so it's good to give more openings :)

djyou replied:
I suggest if if we switch to docker system info or docker info when we need the extra info, so right now if docker -v has the info you need, maybe better to stick with docker -v

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,420 @@
+    accumulated_errors = 0
+
+    #ACR CLI check
+    output = getoutput('az --version')
```
can you reuse the same code that gets acr version, you are already in az cli

ni-reis replied:
which code is that? Are you referring to the code that is invoked by `az --version`?

ni-reis replied:
I hardcoded the version, as I didn't find out how azure cli retrieves the version from each module. Although this is an option, it's not a good idea as we would have to change this every time we change the CLI and we might end up forgetting to update it. The idea is to find the code that gets acr version and adapt it

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,420 @@
+    for line in lines:
+        if line.startswith('Client'):
+            try:
+                client_version = regex.search(line).group(1)
```
Do you plan to validate docker/helm versions such as a minimum required version in the future? Otherwise, it doesn't seem to be necessary to parse the result here either, can we just print the two lines of version info?

Another thing is that docker/helm client changes can easily break you for this kind of context search.

ni-reis replied:
Yes, this is a future planning. Also, doesn't this go against the idea of making the output consistent?
I agree that this can get obsoleted if there are changes on the client display, but maybe we can incorporate a lot of regex to cover multiple versions

djyou replied:
I am not fully sure with such context parsing. If you don't plan to do anything with the version number, do we still need extract it and come up with a new message? Can we just print the raw string as is?

Your code can be easily broken if helm (or docker, for that matter) makes a change to print a different string.

ni-reis replied:
I see. Makes sense, sounds the best idea for now.

ni-reis replied:
Adding this in next push.

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,420 @@
+
+    # Check access to login endpoint
+    url = 'https://' + login_server + '/v2/'
+    challenge = requests.get(url)
```
I would suggest some refactoring with existing methods in _docker_utils since we already of the following logic in place.

This line will actually throw if you couldn't connect, see https://github.com/Azure/azure-cli/blob/70a9d7fda754267a20f8593adf9f2013ae18f74f/src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_docker_utils.py#L179

djyou replied:
Another thing I was thinking was, should we do a forward check (i.e., starting from basic ones) or a backward check (i.e., starting from challenge, if failed, run more basic ones, etc.).

ni-reis replied:
Actually we are starting from basic ones (sequence is first make a DNS lookup to login server, challenge endpoint and then refresh/access token). As those tests are sequential, we only proceed to a test if all before tests succeeded. this line would not throw since it's only called if DNS lookup to login server succeeded

djyou replied:
Do you think a backward check is more efficient? such as if challenge succeeded with 401, do we still need to check dns, etc.?

ni-reis replied:
Considering the usual scenario where the registry login server is correct, I think that backwards check is indeed more efficient (since it makes one less http request), but from the dev side I think it makes more sense to do a forward check. I'm not sure with which one should we proceed though

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,428 @@
+    print(Fore.RED + str(message) + Style.RESET_ALL)
+
+
+def _error_handler(error, operation, ignore_errors, additional_message=None):
```
_handle_error?

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,428 @@
+def _raise_exception(accumulated_errors, error_message=None):
+    if accumulated_errors == 0:
+        print("All checks passed. Please refer to https://aka.ms/acr/faq for more informations.")
+        return None
```
Just return?

ni-reis replied:
Every command must return an object or None. Since this is the last method called, we should return None

ni-reis replied:
I will change the name of this method to "_handle_exception" as we don't always raise the exception.

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,428 @@
+    url = 'https://' + login_server + '/v2/'
+    challenge = requests.get(url)
+
+    if challenge.status_code == 403:
```
The following code has a lot of duplication with https://github.com/Azure/azure-cli/blob/f8cae0185ac413c5351804409b829e4ec1ae560f/src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_docker_utils.py#L44

Could you try to see if you can reuse the existing code? Your logic can be added to existing code too to provide additional error messages.

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,428 @@
+def _check_health_connectivity(cmd, registry_name, ignore_errors):
+    accumulated_errors = 0
+
+    if registry_name is None:
```
if not registry_name

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,428 @@
+        from ._docker_utils import get_login_server_suffix
+        suffix = get_login_server_suffix(cmd.cli_ctx)
+        if suffix is None:
+            suffix = '.azurecr.io'
```
this is incorrect for national cloud. Again, please try to reuse existing code, feel free to refactor/improve them.

ni-reis replied:
Yes, I know. We try to retrieve the suffix according to the current cloud using `suffix = get_login_server_suffix(cmd.cli_ctx)`. In case we can't retrieve, I think that we should use the most common one.

djyou replied:
I wouldn't make that assumption. If you couldn't retrieve, you should just error out. But again, this duplicates a bunch of existing code, so please refactor.

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,428 @@
+    except socket.gaierror:
+        pass
+
+    if registry_ip == "":
```
if not registry_ip

ni-reis replied:
acked.

yugangw-msft committed:
release: update versions (#9418)

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,428 @@
+
+def _pull_docker_image(skip_confirmation, ignore_errors):
+    accumulated_errors = 0
+    image = "mcr.microsoft.com/mcr/hello-world"
```
make this a const

ni-reis replied:
acked.

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,428 @@
+
+    if not skip_confirmation:
+        from knack.prompting import prompt_y_n
+        confirmation = prompt_y_n("This will pull the image {}:{}. Proceed?".format(image, tag))
```
There doesn't seem to be any point to split `image` and `tag` since you always combine them. Also image = repo:tag, tag is part of image.

djyou replied:
If you answer no, wouldn't the command exit immediately so you won't be able to execute the below lines?

ni-reis replied:
Makes sense, using image:tag together. If you answer no, the command won't pull the image, but other checks will still happen.

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,428 @@
+    try:
+        docker_command = get_docker_command()
+    except CLIError:
+        accumulated_errors |= _error_handler(Errors.DOCKER_COMMAND_ERROR,
```
There could be different errors thrown from `get_docker_command` such as docker is not installed, docker daemon is not running, etc. Why not print the raw error message to provide more context of what went wrong?

adewaleo committed:
[vm] Fix bug introduced in #9306.  (#9421)

anusapan committed:
Add support for Manual-Failover (#9413)

* add new command

* add support for manual-failover

* update help content

* re-record test session

* Bump SDK version

adewaleo committed:
[core] Fix bug where --query cannot be used with --output yaml (#9403)

* [core] Fix bug where --query cannot be used with --output yaml

* Add comment explaining yaml formatter error handling logic.

ganga1980 committed:
update default log analytics workspace region mapping for aks monitoring addon (#9434)

* add cluster spn with metric publisher role to aks cluster resource

* refactor the code

* fix default workspace issue in china cloud

* add metrics publisher role assignment only in public cloud

* fix pr feedback

* fix build error

* fix lint error

* update region mapping

* order regions in sortorder

* fix trailing whitespaces

marstr committed:
Fix JP installer line (#9438)

Curl must follow redirects. Fixes #9435

julienstroheker committed:
ACS - OSA - Whitelist message (#9409)

* add friendly msg when sub is not whitelisted for GA API

* bump version and add history

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py:
```diff
@@ -214,3 +214,8 @@ def load_arguments(self, _):  # pylint: disable=too-many-statements
         c.argument('ip_address', help='IPv4 address or CIDR range.')
+
+    with self.argument_context('acr check-health') as c:
+        c.argument('registry_name', options_list=['-n', '--name'], help='Name of target registry.', required=False)
```
No need to declare again, the default one should just work https://github.com/Azure/azure-cli/blob/6868f1f8d8e4b30391426908dcca75f7188d2e7e/src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py#L45

ni-reis replied:
Nice catch, done :)

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py:
```diff
@@ -214,3 +214,8 @@ def load_arguments(self, _):  # pylint: disable=too-many-statements
+
+    with self.argument_context('acr check-health') as c:
+        c.argument('registry_name', options_list=['-n', '--name'], help='Name of target registry.', required=False)
+        c.argument('skip_confirmation', options_list=['-y'], help='Skips confirmation, allowing to automatically pull the sample image', action='store_true', required=False)
```
Also use the default one https://github.com/Azure/azure-cli/blob/6868f1f8d8e4b30391426908dcca75f7188d2e7e/src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py#L53

ni-reis replied:
Nice catch! Done :D

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_params.py:
```diff
@@ -214,3 +214,8 @@ def load_arguments(self, _):  # pylint: disable=too-many-statements
+    with self.argument_context('acr check-health') as c:
+        c.argument('registry_name', options_list=['-n', '--name'], help='Name of target registry.', required=False)
+        c.argument('skip_confirmation', options_list=['-y'], help='Skips confirmation, allowing to automatically pull the sample image', action='store_true', required=False)
+        c.argument('ignore_errors', options_list=['--ignore'], help='Ignore errors, displaying them only at the final', action='store_true', required=False)
```
I have a minor pick on the param `--ignore` (i.e., ignore what?), the original `--ignore-errors` might be better, unless there is an established convention such as `--yes` that everyone knows.

ni-reis replied:
Agreed, moved to `--ignore-errors`.

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/helm.py:
```diff
@@ -187,12 +187,16 @@ def _get_helm_command():
-        raise CLIError(stderr.decode())
+        if is_diagnostics_context:
+            return None, UNEXPECTED_ERROR.append_error_message(stderr.decode())
+        raise CLIError(UNEXPECTED_ERROR.append_error_message(stderr.decode()).get_error_message())
```
Can we combine get_error_message and append_error_message so we only have one call here?

ni-reis replied:
I think it's hard, the idea for `append_error_message` is to return a ErrorClass model with the appended error message. We use this in other methods that deals with ErrorClass objects, not with messages, which make it difficult to combine both.

ni-reis replied:
The problem is that, in diagnostics context, we want to deal with a `ErrorClass` object, while raising a CLI error we want a string (obtained by `get_error_message`.

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/custom.py:
```diff
@@ -242,18 +241,26 @@ def _get_docker_command():
             logger.debug("Could not run '%s' command. Exception: %s", docker_command, str(inner))
-            raise CLIError(docker_not_available)
+            if is_diagnostics_context:
+                return docker_command, DOCKER_DAEMON_ERROR
```
I think you can always return None, Error, it also avoids anyone accidentally use "docker" when there is an error

ni-reis replied:
The problem with it is that we can run `docker --version` even without docker daemon, so my idea was to return None only if the error was with the command, not with docker daemon. Do you think we should just disregard everything anyways? It might also be a good option...

djyou commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,259 @@
+
+    # Check access to login endpoint
+    url = 'https://' + login_server + '/v2/'
+    result_from_token = _get_aad_token(cmd.cli_ctx, login_server, False, is_diagnostics_context=True)
```
Can you check if we can refactor and use this method https://github.com/Azure/azure-cli/blob/6868f1f8d8e4b30391426908dcca75f7188d2e7e/src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/_docker_utils.py#L134 since there is some other checks such as vnet, admin-enabled, etc. so you can also avoid handling suffix here.

ni-reis replied:
I was thinking of refactoring this method later, maybe even created a method just to check the dns lookup to the registry and the VNET status and incorporate this in both methods (this one and check health method). The reason for this is because this method as is more specific for logging in. Also we don't want to include checks of admin status on diagnosis.

yugangw-msft committed:
auth: handle unicode in subscription name when under python 2.7 (#9439)

yugangw-msft committed:
role: update the help of --assignee-object-id (#9457)

* role: update the help of --assignee-object-id

* fix grammar

bgklein committed:
Update error logging for reported unclear issues (#9180)

* Update error logging for reported unclear issues

* Combine error messages

* fix ci

zhedahht committed:
Enable/Disable AKS kube-dashboard addon (#9458)

* Disable/Enable kube-dashboard addon

* Minor fixes

* change HISTORY.rst

Commit:
Added ACR command to create a buildpack task (#9366)

* Added acr pack command scaffold

* Added acr pack test

* Fixed command registration

* Consolidated value

* Added command module scaffold

* Extracted method

* Added linux assertion

* Fixed params

* Moved validation to be earlier

No need to upload sources if the build cannot proceed

* Removed --file, added --image

* Improved docstring

* Added builder flag

* Fixed flake9 errors in run.py

* Fixed flake8 erros in pack.py

* Added registry name env var to build

* Removed test

* Unused

* CR remark

* Merge remote-tracking branch 'refs/remotes/Azure/dev' into dev

# Conflicts:
#	src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/run.py

* Added task version

* Removed unnecessary doc

* Removed unused imports

* Added no-pull arg

* Flipped arg

* Added automatic registry prefix

* CR

Commit:
changing history.rst and fixing helm command

Commit:
fixing flagged checks

marstr committed:
Make the RPM Spec flexible to EL7 vs EL8 (#9394)

* Add python3 macro to RPM spec

* fixing missed raw python use

* fixing another missed raw python use

* Don't look for el7, look for el*

* Try removing duplicate dependencies in Runtime and buildtime requirements

* Declare dependency on virtualenv instead of downloading it

* Disable python-devel requirement

* WIP

* More fixes

* Exclude hardlinks created by virtualenv

In reality, all we need from virtualenv is an easy way to build up an alternate site-packages directory. Infact, what this is doing just emulates how a virtualenv works, but without the hardlinks that were confusing rpmbuild!

* Ignore user site-packages

djyou committed:
[ACR] Return patch response directly for acr repository update (#9420)

* Return patch response directly for acr repository update

* fix ci

* Remove deprecated classic registry error handling

avtakkar commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/custom.py:
```diff
@@ -242,18 +241,26 @@ def _get_docker_command():
+            return docker_command, DOCKER_DAEMON_ERROR
+        raise CLIError(DOCKER_DAEMON_ERROR.get_error_message())
 
     if stderr:
```
`docker ps` will write a warning to stderr if `.docker/config.json` requires elevated access. For example, if there are secrets in that file. In that case, throwing an error is probably not necessary. We could ignore the warning and carry on.

avtakkar commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,266 @@
+# Get current CLI version
+def _get_cli_version():
+
+    acr_cli_version = "2.2.8"
```
Read the version from `src/command_modules/azure-cli-acr/setup.py` for easy maintenance.

marstr committed:
Fixing Dockerfile name of RPM builder for centos in pipeline definition (#9493)

yugangw-msft committed:
test: fix a flaky test (#9494)

avtakkar commented on src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/check_health.py:
```diff
@@ -0,0 +1,266 @@
+
+    # Docker version check
+    output, stderr = _subprocess_communicate([docker_command, "--version"])
+    if stderr:
```
Same here. Some warnings written to stderr regarding config file loading can potentially be ignored. Such as: `WARNING: Error loading config file: /xyz/.docker/config.json: stat /xyz/.docker/config.json: permission denied`

macedogm committed:
Update _help.py (#9502)

Corrected a small typo.

jdeschenes committed:
Added Error Message when no TTY is available (#9481)

* Added Error Message when no TTY is available

Fix #3876

When performing a resource group deployment, if no TTY is available,
a CLIError will be raised detailing the parameters that were not
provided.

* Fixes for python2 compatibility

tjprescott committed:
[Network] AppGateway WAF Policy commands (#9470)

* initial work.

* WAF Policy progress.

* Progress.

* WAF policy commands and test.

* Fix and re-record test.

* Closes #9255.

* Fix test issue.

* Code review feedback.

* Fix update logic.

* Fix CI

adewaleo committed:
[vm] create now more intelligently uses image data disk lun info. (#9504)

* vm create now more intelligently uses image data disk lun information.

* Re-record and update a test. Other minor updates.

* Update tests. Minor bug fixes.

* Upate hybrid profile tests

dorfire committed:
Fixed registry prefix check (#9509)

* Added acr pack command scaffold

* Added acr pack test

* Fixed command registration

* Consolidated value

* Added command module scaffold

* Extracted method

* Added linux assertion

* Fixed params

* Moved validation to be earlier

No need to upload sources if the build cannot proceed

* Removed --file, added --image

* Improved docstring

* Added builder flag

* Fixed flake9 errors in run.py

* Fixed flake8 erros in pack.py

* Added registry name env var to build

* Removed test

* Unused

* CR remark

* Merge remote-tracking branch 'refs/remotes/Azure/dev' into dev

# Conflicts:
#	src/command_modules/azure-cli-acr/azure/cli/command_modules/acr/run.py

* Added task version

* Removed unnecessary doc

* Removed unused imports

* Added no-pull arg

* Flipped arg

* Added automatic registry prefix

* CR

* Fixed registry prefix

* Updated history.rst

tjprescott committed:
Update CLI to use Knack 0.6.2 (preview mechanism) (#9456)

* Initial work.

* Update preview references.

* Additional work to wire up preview arguments.

* Update for knack changes.

* Update dependencies. Remove command suggestion code. Deprecate BatchAI. Hide ACS.

* Restore CLI-specific spellchecker.

* Update history entries.

tjprescott committed:
Network API 2019-04-01 (#9523)

* Re-record tests for 2019-04-01.

* Closes #9463.

* Closes #9401. Restore private endpoint tests and commands.

aim-for-better committed:
support ssh_public_key (#9512)

tjprescott committed:
Fix #9489. (#9531)

tjprescott committed:
Fix #9443. (#9533)

Commit:
ServiceBus: Fix for Issue 9319 (#9538)

* Fix for Issue - 9319

* updated version in histroy and setup files

* updated code

Commit:
Fixing some comments and Travis CI issues

Commit:
Improving history and fixing version count

Commit:
That one blank line

Commit:
Improving history message and fixing acr cli version
